File: CyxbsMobile2019_iOS/Rising Schedule Custom/ScheduleCustomEditView.m
//
//  ScheduleCustomEditView.m
//  CyxbsMobile2019_iOS
//
//  Created by SSR on 2023/2/28.
//  Copyright © 2023 Redrock. All rights reserved.
//

#import "ScheduleCustomEditView.h"

#import "ScheduleCustomEditCollectionViewCell.h"

#import "ScheduleNeedsSupport.h"

#pragma mark - ScheduleCustomEditView ()

@interface ScheduleCustomEditView () <
    UICollectionViewDelegateFlowLayout,
    UICollectionViewDataSource,

    UIPickerViewDelegate,
    UIPickerViewDataSource,

    UIGestureRecognizerDelegate,

    UITextFieldDelegate
>

@property (nonatomic, strong) UILabel *topLab;

@property (nonatomic, strong) UILabel *titleLab;
@property (nonatomic, strong) UITextField *titleTextField;

@property (nonatomic, strong) UILabel *contentLab;
@property (nonatomic, strong) UITextField *contentTextField;

@property (nonatomic, strong) UILabel *sectionLab;
@property (nonatomic, strong) UICollectionView *collectionView;

@property (nonatomic, strong) UILabel *periodLab;
@property (nonatomic, strong) UIPickerView *periodPicker;

@property (nonatomic, strong) UIButton *finBtn;

@property (nonatomic, copy) NSIndexPath *l_idx;

@end

#pragma mark - ScheduleCustomEditView

@implementation ScheduleCustomEditView {
    NSMutableIndexSet *_mutiIdxSet;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = UIColor.clearColor;
        [self addSubview:self.topLab];
        
        [self addSubview:self.titleLab];
        [self addSubview:self.titleTextField];
        
        [self addSubview:self.contentLab];
        [self addSubview:self.contentTextField];
        
        [self addSubview:self.sectionLab];
        [self addSubview:self.collectionView];
        
        [self addSubview:self.periodLab];
        [self addSubview:self.periodPicker];
        
        [self addSubview:self.finBtn];
        
        UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(_endEdit:)];
        tap.delegate = self;
        [self addGestureRecognizer:tap];
        
        _period = NSMakeRange(1, 1);
        _mutiIdxSet = NSMutableIndexSet.indexSet;
    }
    return self;
}

#pragma mark - Method

// title

- (void)setTitle:(NSString *)title {
    if ([title isEqualToString:@""]) {
        title = nil;
    }
    self.titleTextField.text = title;
}

- (NSString *)title {
    return self.titleTextField.text.copy;
}

// content

- (void)setContent:(NSString *)content {
    if ([content isEqualToString:@""]) {
        content = nil;
    }
    self.contentTextField.text = content;
}

- (NSString *)content {
    return self.contentTextField.text.copy;
}

// sections

- (void)setSections:(NSIndexSet *)sections {
    _mutiIdxSet = sections ? sections.mutableCopy : NSMutableIndexSet.indexSet;
    if (_collectionView) {
        [sections enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL * __unused stop) {
            NSIndexPath *idxPath = [NSIndexPath indexPathForItem:(idx - 1) % 8 inSection:(idx - 1) / 8];
            [self.collectionView selectItemAtIndexPath:idxPath animated:YES scrollPosition:UICollectionViewScrollPositionLeft];
        }];
    }
}

- (NSIndexSet *)sections {
    return _mutiIdxSet.copy;
}

// inWeek

- (void)setInWeek:(NSInteger)inWeek {
    if (inWeek < 1 || inWeek > 7) {
        inWeek = 0;
    }
    [self.periodPicker selectRow:inWeek - 1 inComponent:0 animated:YES];
}

- (NSInteger)inWeek {
    return [self.periodPicker selectedRowInComponent:0] + 1;
}

// period

- (void)setPeriod:(NSRange)period {
    if (period.location < 1 || period.location > 12) {
        period.location = 1;
        period.length = 1;
    }
    _period = period;
    [self.periodPicker selectRow:period.location - 1 inComponent:1 animated:YES];
    [self.periodPicker reloadComponent:2];
    [self.periodPicker selectRow:period.length - 1 inComponent:2 animated:YES];
}

#pragma mark - Lazy

- (UILabel *)topLab {
    if (_topLab == nil) {
        _topLab = self._lab;
        _topLab.text = @"为你的行程添加";
        [_topLab sizeToFit];
        _topLab.left = 18;
        _topLab.top = 0;
    }
    return _topLab;
}

- (UILabel *)titleLab {
    if (_titleLab == nil) {
        _titleLab = self._lab;
        _titleLab.text = @"一个标题";
        [_titleLab sizeToFit];
        _titleLab.left = self.topLab.left;
        _titleLab.top = self.topLab.bottom;
    }
    return _titleLab;
}

- (UITextField *)titleTextField {
    if (_titleTextField == nil) {
        _titleTextField = self._textField;
        _titleTextField.frame = CGRectMake(self.titleLab.left, self.titleLab.bottom + 5, self.width - 2 * 16, 50);
        _titleTextField.placeholder = @"标题";
    }
    return _titleTextField;
}

- (UILabel *)contentLab {
    if (_contentLab == nil) {
        _contentLab = self._lab;
        _contentLab.text = @"具体内容";
        [_contentLab sizeToFit];
        _contentLab.left = self.titleTextField.left;
        _contentLab.top = self.titleTextField.bottom + 20;
    }
    return _contentLab;
}

- (UITextField *)contentTextField {
    if (_contentTextField == nil) {
        _contentTextField = self._textField;
        _contentTextField.frame = CGRectMake(self.contentLab.left, self.contentLab.bottom + 5, self.width - 2 * 16, 50);
        _contentTextField.placeholder = @"内容";
    }
    return _contentTextField;
}

- (UILabel *)sectionLab {
    if (_sectionLab == nil) {
        _sectionLab = self._lab;
        _sectionLab.text = @"选择周数";
        [_sectionLab sizeToFit];
        _sectionLab.left = self.contentTextField.left;
        _sectionLab.top = self.contentTextField.bottom + 20;
    }
    return _sectionLab;
}

- (UICollectionView *)collectionView {
    if (_collectionView == nil) {
        UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
        layout.minimumLineSpacing = 2;
        layout.minimumInteritemSpacing = 2;
        layout.sectionInset = UIEdgeInsetsMake(3, 0, 3, 0);
        
        CGFloat left = self.sectionLab.left;
        _collectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(left, self.sectionLab.bottom + 5, self.width - 2 * left, 105) collectionViewLayout:layout];
        _collectionView.backgroundColor = UIColor.clearColor;
        _collectionView.dataSource = self;
        _collectionView.delegate = self;
        _collectionView.showsVerticalScrollIndicator = NO;
        _collectionView.showsHorizontalScrollIndicator = NO;
        _collectionView.bounces = NO;
        _collectionView.allowsMultipleSelection = YES;
        [_collectionView registerClass:ScheduleCustomEditCollectionViewCell.class forCellWithReuseIdentifier:ScheduleCustomEditCollectionViewCellReuseIdentifier];
    }
    return _collectionView;
}

- (UILabel *)periodLab {
    if (_periodLab == nil) {
        _periodLab = self._lab;
        _periodLab.text = @"确定时间";
        [_periodLab sizeToFit];
        _periodLab.top = self.collectionView.bottom + 15;
        _periodLab.left = self.collectionView.left;
    }
    return _periodLab;
}

- (UIPickerView *)periodPicker {
    if (_periodPicker == nil) {
        CGFloat left = self.periodLab.left;
        _periodPicker = [[UIPickerView alloc] initWithFrame:CGRectMake(left, self.periodLab.bottom - 20, self.width - 2 * left, 216)];
        _periodPicker.delegate = self;
        _periodPicker.dataSource = self;
    }
    return _periodPicker;
}

- (UIButton *)finBtn {
    if (_finBtn == nil) {
        _finBtn = [[UIButton alloc] initWithFrame:CGRectMake(0, 0, 66, 66)];
        _finBtn.centerX = self.width / 2;
        _finBtn.centerY = (self.periodPicker.bottom + self.height) / 2;
        _finBtn.layer.cornerRadius = 20;
        _finBtn.backgroundColor = UIColorHex(#AABBFF);
        _finBtn.enabled = NO;
        
        [_finBtn setImage:[UIImage imageNamed:@"nextpage"] forState:UIControlStateNormal];
        _finBtn.imageEdgeInsets = UIEdgeInsetsMake(20, 20, 20, 20);
        
        [_finBtn addTarget:self action:@selector(_tapBtn:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _finBtn;
}

#pragma mark - Private

- (void)_endEdit:(UITapGestureRecognizer *)tap {
    if (tap.state == UIGestureRecognizerStateEnded) {
        [self endEditing:YES];
        [self _checkBtn];
    }
}

- (void)_tapBtn:(UIButton *)btn {
    if (self.delegate && [self.delegate respondsToSelector:@selector(scheduleCustomEditViewDidFinishEditing:)]) {
        [self.delegate scheduleCustomEditViewDidFinishEditing:self];
    }
}

- (void)_checkBtn {
    if (self.title && self.content && _mutiIdxSet.count != 0) {
        self.finBtn.enabled = YES;
        self.finBtn.backgroundColor = UIColorHex(#4841E2);
    } else {
        self.finBtn.enabled = NO;
        self.finBtn.backgroundColor = UIColorHex(#AABBFF);
    }
}

#pragma mark - <UITextFieldDelegate>

- (void)textFieldDidEndEditing:(UITextField *)textField {
    [self _checkBtn];
}

#pragma mark - <UIGestureRecognizerDelegate>

- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer {
    CGPoint point = [gestureRecognizer locationInView:self];
    if (CGRectContainsPoint(self.collectionView.frame, point)) {
        return NO;
    }
    if (CGRectContainsPoint(self.finBtn.frame, point)) {
        return NO;
    }
    return YES;
}

#pragma mark - <UICollectionViewDelegate>

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    [_mutiIdxSet addIndex:indexPath.section * 8 + indexPath.item + 1];
    [self _checkBtn];
}

- (void)collectionView:(UICollectionView *)collectionView didDeselectItemAtIndexPath:(NSIndexPath *)indexPath {
    [_mutiIdxSet removeIndex:indexPath.section * 8 + indexPath.item + 1];
    [self _checkBtn];
}

#pragma mark - <UICollectionViewDelegateFlowLayout>

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    CGFloat width = collectionView.width / 8 - 2;
    return CGSizeMake(width, width * 0.62);
}

#pragma mark - <UICollectionViewDataSource>

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 3;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section  {
    return 8;
}

- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    ScheduleCustomEditCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:ScheduleCustomEditCollectionViewCellReuseIdentifier forIndexPath:indexPath];
    
    NSUInteger week = indexPath.section * 8 + indexPath.item + 1;
    cell.title = [NSString stringWithFormat:@"第%ld周", week];
    
    return cell;
}

#pragma mark - <UIPickerViewDataSource>

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 3;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    if (component == 0) {
        return 7;
    }
    if (component == 1) {
        return 12;
    }
    if (component == 2) {
        return 13 - self.period.location;
    }
    return 0;
}

#pragma mark - <UIPickerViewDelegate>

- (NSAttributedString *)pickerView:(UIPickerView *)pickerView attributedTitleForRow:(NSInteger)row forComponent:(NSInteger)component {
    NSString *str = @"";
    if (component == 0) {
        switch (row) {
            case 0: str = @"星期一"; break;
            case 1: str = @"星期二"; break;
            case 2: str = @"星期三"; break;
            case 3: str = @"星期四"; break;
            case 4: str = @"星期五"; break;
            case 5: str = @"星期六"; break;
            case 6: str = @"星期天"; break;
            default: return nil;
        }
    }
    if (component == 1) {
        str = [NSString stringWithFormat:@"从第%ld节", row + 1];
    }
    if (component == 2) {
        str = [NSString stringWithFormat:@"到第%ld节", self.period.location + row];
    }
    return [[NSAttributedString alloc] initWithString:str attributes:@{
        NSFontAttributeName: [UIFont fontWithName:FontName.PingFangSC.Medium size:18],
        NSForegroundColorAttributeName: [UIColor Light:UIColorHex(#15315B) Dark:UIColorHex(#F0F0F2)]
    }];
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    if (component == 1) {
        _period = NSMakeRange(row + 1, [pickerView selectedRowInComponent:2] + 1);
        [pickerView reloadComponent:2];
    }
    if (component == 2) {
        _period.length = [pickerView selectedRowInComponent:2] + 1;
    }
}

#pragma mark - Factory

- (UILabel *)_lab {
    UILabel *lab = [[UILabel alloc] init];
    lab.font = [UIFont fontWithName:FontName.PingFangSC.Semibold size:28];
    lab.textColor = [UIColor Light:UIColorHex(#122D55) Dark:UIColorHex(#F0F0F2)];
    return lab;
}

- (UITextField *)_textField {
    UITextField *_tf = [[UITextField alloc] init];
    _tf.leftView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 17, 0)];
    _tf.leftViewMode = UITextFieldViewModeAlways;
    _tf.font = [UIFont fontWithName:FontName.PingFangSC.Semibold size:18];
    _tf.textColor = [UIColor Light:UIColorHex(#15315B) Dark:UIColorHex(#F0F0F2)];
    _tf.backgroundColor = [UIColor Light:UIColorHex(#F2F3F7) Dark:UIColorHex(#2D2D2D)];
    _tf.layer.cornerRadius = 17;
    return _tf;
}

@end

File: CyxbsMobile2019_iOS/Curriculum/Edit/View/ScheduleCustomEditView.swift
//
//  ScheduleCustomEditView.swift
//  CyxbsMobile2019_iOS
//
//  Created by SSR on 2023/10/9.
//  Copyright © 2023 Redrock. All rights reserved.
//

import UIKit

class ScheduleCustomEditView: UIView {
    
    var backBtnTaped: ((ScheduleCustomEditView) -> ())?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        isUserInteractionEnabled = true
        addSubview(promptTitleLab)
        addSubview(titleLab)
        addSubview(titleTextField)
        addSubview(placeLab)
        addSubview(placeTextField)
        addSubview(sectionLab)
        addSubview(sectionCollectionView)
        addSubview(periodLab)
        addSubview(periodPicker)
        addSubview(finishBtn)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func sizeThatFits(_ size: CGSize) -> CGSize {
        var size = size
        size.height = finishBtn.frame.maxY + sectionSpace
        return size
    }
    
    // MARK: lazy
    
    lazy var promptTitleLab: UILabel = {
        let lab = createTitleStyleLab(title: "为你的行程添加")
        lab.frame.origin = CGPoint(x: 18, y: itemSpace)
        return lab
    }()
    
    lazy var titleLab: UILabel = {
        let lab = createTitleStyleLab(title: "一个标题")
        lab.frame.origin = CGPoint(x: promptTitleLab.frame.minX, y: promptTitleLab.frame.maxY)
        return lab
    }()

    lazy var titleTextField: UITextField = {
        let textField = createTextField(placeholder: "例如：自习")
        textField.frame.origin.y = titleLab.frame.maxY + itemSpace
        return textField
    }()
    
    lazy var placeLab: UILabel = {
        let lab = createTitleStyleLab(title: "具体内容")
        lab.frame.origin = CGPoint(x: titleLab.frame.minX, y: titleTextField.frame.maxY + sectionSpace)
        return lab
    }()
    
    lazy var placeTextField: UITextField = {
        let textField = createTextField(placeholder: "例如：红岩网校工作站")
        textField.frame.origin.y = placeLab.frame.maxY + itemSpace
        return textField
    }()
    
    lazy var sectionLab: UILabel = {
        let lab = createTitleStyleLab(title: "选择周数")
        lab.frame.origin = CGPoint(x: placeLab.frame.minX, y: placeTextField.frame.maxY + sectionSpace)
        return lab
    }()
    
    lazy var sectionCollectionView: UICollectionView = {
        let width = bounds.width
        let layout = UICollectionViewFlowLayout()
        layout.scrollDirection = .vertical
        layout.minimumLineSpacing = 2
        layout.minimumInteritemSpacing = 2
        layout.sectionInset = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
        let itemWidth = (width - layout.sectionInset.left - layout.sectionInset.right) / 6 - layout.minimumInteritemSpacing
        layout.itemSize = CGSize(width: itemWidth, height: itemWidth * 0.62)
        
        let collectionView = UICollectionView(frame: CGRect(x: 0, y: sectionLab.frame.maxY + itemSpace, width: width, height: layout.itemSize.height * 4 + layout.minimumLineSpacing * 3), collectionViewLayout: layout)
        collectionView.backgroundColor = .clear
        collectionView.delegate = self
        collectionView.dataSource = self
        collectionView.showsHorizontalScrollIndicator = false
        collectionView.showsVerticalScrollIndicator = false
        collectionView.bounces = false
        collectionView.allowsMultipleSelection = true
        collectionView.register(ScheduleEditSectionCollectionViewCell.self, forCellWithReuseIdentifier: ScheduleEditSectionCollectionViewCell.identifier)
        return collectionView
    }()
    
    lazy var periodLab: UILabel = {
        let lab = createTitleStyleLab(title: "确定时间")
        lab.frame.origin = CGPoint(x: sectionLab.frame.minX, y: sectionCollectionView.frame.maxY + sectionSpace)
        return lab
    }()
    
    lazy var periodPicker: UIPickerView = {
        let pickerView = UIPickerView(frame: CGRect(x: 0, y: periodLab.frame.maxY, width: bounds.width, height: 216))
        pickerView.delegate = self
        pickerView.dataSource = self
        return pickerView
    }()
    
    lazy var finishBtn: UIButton = {
        let btn = UIButton()
        btn.isEnabled = false
        btn.frame.origin.y = periodPicker.frame.maxY + sectionSpace
        btn.frame.size = CGSize(width: 66, height: 66)
        btn.center.x = bounds.width / 2
        btn.layer.cornerRadius = 20
        let image = UIImage(named: "direction_goto_right")?
            .scaled(toHeight: 20)?
            .withRenderingMode(.alwaysOriginal)
        btn.setImage(image, for: .normal)
        btn.addTarget(self, action: #selector(touchUpInside(fishishBtn:)), for: .touchUpInside)
        btn.backgroundColor = .hex("#AABBFF")
        return btn
    }()
}

// MARK: creator

extension ScheduleCustomEditView {
    
    var itemSpace: CGFloat { 10 }
    
    var sectionSpace: CGFloat { 20 }
    
    func createTitleStyleLab(title: String?) -> UILabel {
        let lab = UILabel()
        lab.font = .systemFont(ofSize: 28, weight: .semibold)
        lab.textColor = .ry(light: "#122D55", dark: "#F0F0F2")
        lab.text = title
        lab.sizeToFit()
        return lab
    }
    
    func createTextField(placeholder: String?, leftSpace: CGFloat = 16) -> UITextField {
        let textField = UITextField(frame: CGRect(x: leftSpace, y: 0, width: bounds.width - 2 * leftSpace, height: 50))
        textField.layer.cornerRadius = 15
        textField.backgroundColor = .ry(light: "#F2F3F7", dark: "#2D2D2D")
        textField.font = .systemFont(ofSize: 18, weight: .semibold)
        textField.textColor = .ry(light: "#15315B", dark: "#F0F0F2")
        textField.placeholder = placeholder
        textField.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 17, height: 0))
        textField.leftViewMode = .always
        textField.delegate = self
        return textField
    }
}

// MARK: useable

extension ScheduleCustomEditView {
    
    var title: String {
        titleTextField.text ?? ""
    }
    
    var place: String {
        placeTextField.text ?? ""
    }
    
    var sections: IndexSet {
        IndexSet(
            sectionCollectionView.indexPathsForSelectedItems?.map { $0.item + 1 } ?? []
        )
    }
    
    var inWeek: Int {
        periodPicker.selectedRow(inComponent: 0) + 1
    }
    
    var period: ClosedRange<Int> {
        (periodPicker.selectedRow(inComponent: 1) + 1) ...
        (periodPicker.selectedRow(inComponent: 1) + periodPicker.selectedRow(inComponent: 2) + 1)
    }
}

// MARK: interactive

extension ScheduleCustomEditView {
    
    func prepare(title: String? = nil, place: String? = nil, sections: IndexSet = [], inWeek: Int, location: Int, lenth: Int = 1) {
        titleTextField.text = title
        placeTextField.text = place
        for section in sections {
            sectionCollectionView.selectItem(at: IndexPath(item: max(0, section - 1), section: 0), animated: true, scrollPosition: .right)
        }
        let inWeek = min(max(1, inWeek), 7)
        periodPicker.selectRow(inWeek - 1, inComponent: 0, animated: true)
        let location = min(max(1, location), 12)
        periodPicker.selectRow(location - 1, inComponent: 1, animated: true)
        periodPicker.reloadComponent(2)
        let lenth = min(max(1, lenth), periodPicker.numberOfRows(inComponent: 2))
        periodPicker.selectRow(lenth - 1, inComponent: 2, animated: true)
    }
    
    var isInfoVisable: Bool {
        title.count > 0 &&
        place.count > 0 &&
        sections.count > 0
    }
    
    func checkInfoVisable() {
        finishBtn.isEnabled = isInfoVisable
        if isInfoVisable {
            finishBtn.backgroundColor = .hex("#4841E2")
        } else {
            finishBtn.backgroundColor = .hex("#AABBFF")
        }
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        endEditing(true)
    }
    
    @objc
    func touchUpInside(fishishBtn: UIButton) {
        backBtnTaped?(self)
    }
}

// MARK: UITextFieldDelegate

extension ScheduleCustomEditView: UITextFieldDelegate {
    
    func textFieldDidEndEditing(_ textField: UITextField, reason: UITextField.DidEndEditingReason) {
        checkInfoVisable()
    }
}

// MARK: UICollectionViewDataSource

extension ScheduleCustomEditView: UICollectionViewDataSource {
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        24
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: ScheduleEditSectionCollectionViewCell.identifier, for: indexPath) as! ScheduleEditSectionCollectionViewCell
        
        cell.title = ScheduleDataFetch.sectionString(withSection: indexPath.item + 1)
        
        return cell
    }
}

// MARK: UICollectionViewDelegate

extension ScheduleCustomEditView: UICollectionViewDelegate {
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        checkInfoVisable()
    }
    
    func collectionView(_ collectionView: UICollectionView, didDeselectItemAt indexPath: IndexPath) {
        checkInfoVisable()
    }
}

// MARK: UIPickerViewDataSource

extension ScheduleCustomEditView: UIPickerViewDataSource {
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        3
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        switch component {
        case 0: return 7
        case 1: return 12
        case 2: return 12 - pickerView.selectedRow(inComponent: 1)
        default: return 0
        }
    }
}

// MARK: UIPickerViewDelegate

extension ScheduleCustomEditView: UIPickerViewDelegate {
    
    func pickerView(_ pickerView: UIPickerView, attributedTitleForRow row: Int, forComponent component: Int) -> NSAttributedString? {
        var str: String? = nil
        switch component {
        case 0:
            switch row {
            case 0: str = "星期一"
            case 1: str = "星期二"
            case 2: str = "星期三"
            case 3: str = "星期四"
            case 4: str = "星期五"
            case 5: str = "星期六"
            case 6: str = "星期天"
            default: break
            }
        case 1:
            str = "从第\(row + 1)节"
        case 2:
            str = "到第\(pickerView.selectedRow(inComponent: 1) + row + 1)节"
        default: break
        }
        guard let str else { return nil }
        return NSAttributedString(string: str, attributes: [
            .font: UIFont.systemFont(ofSize: 12, weight: .medium),
            .foregroundColor: UIColor.ry(light: "#15315B", dark: "#F0F0F2")
        ])
    }
    
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        if component == 1 {
            pickerView.reloadComponent(2)
        }
    }
}
